doc/async_model.md

```diff
@@ -0,0 +1,198 @@
+Overview of asynchronous programming
+====================================
+
+Let's say we have a program that reads a large file and processes each chunk in the file in some way. A naive implementation would look as following:
+
+    while (chunk = read_chunk()) {
+        process_chunk(chunk);
+    }
+
+The problem with the naive implementation is that although I/O and CPU are different system resources, it uses only one each time instead of reading and processing at the same time. Asynchronous programming tries to solve this problem by allowing the read operation to run WITHOUT waiting for its result. This can be done either via asynchronous callbacks or by concurrency. A solution based on asynchronous callbacks would look as following:
+
+    // The function "callback" is called when the chunk is read.
+    auto callback = [](Chunk&& chunk) {
+        if (chunks_remaining_to_read())
+            read_chunk_async(callback);
```

Comment:
I am not sure that recursive lambdas is a valid thing. I think in this case you must use named functions.

Reply:
recursive lambdas work as long as auto is not used, I will edit this accordingly.